File: data/mb_genres_scraper.py
#!/usr/bin/env python3
import json
import re
import sys
import time
from pathlib import Path
from typing import List, Dict, Any

import requests
from bs4 import BeautifulSoup

BASE_URL = "https://musicbrainz.org/genre/"
HEADERS = {"User-Agent": "GenreScraper/1.1 (example@example.com)"}

TARGET_ROWS = {
    "subgenre of:":       "subgenre_of",
    "subgenres:":         "subgenres",
    "has fusion genres:": "fusion_genres",
    "fusion of:":         "fusion_of",
    "influenced by:":     "influenced_by",
    "influenced genres:": "influenced_genres",
}

UUID_RE = re.compile(r"/genre/([0-9a-f\-]{36})", re.I)

def _extract_links(cell) -> List[Dict[str, str]]:
    links = []
    for a in cell.find_all("a", href=True):
        match = UUID_RE.search(a["href"])
        if match:
            genre_id = match.group(1)
            bdi = a.find("bdi")
            name = bdi.get_text(strip=True) if bdi else a.get_text(strip=True)
            links.append({"id": genre_id, "name": name})
    return links


def scrape_single(genre: Dict[str, Any]) -> Dict[str, Any]:
    genre_id = genre["id"]
    name = genre["name"]
    artist_count = genre.get("artistCount", 0)

    url = f"{BASE_URL}{genre_id}"
    try:
        resp = requests.get(url, headers=HEADERS, timeout=30)
        resp.raise_for_status()
    except Exception as e:
        print(f"Error fetching {genre_id} – {name}: {e}", file=sys.stderr)
        return {
            "id": genre_id,
            "name": name,
            "artistCount": artist_count,
            "wikipedia": None,
            **{k: [] for k in TARGET_ROWS.values()}
        }

    soup = BeautifulSoup(resp.text, "html.parser")

    # ---- Wikipedia section ----
    # wikipedia_text = None
    # wiki_section = soup.find("section", class_="wikipedia")
    # if wiki_section:
    #     first_paragraph = wiki_section.find("p")
    #     if first_paragraph:
    #         # Remove inline formatting tags (like <b>, <i>, <a>, etc.)
    #         for tag in first_paragraph.find_all(["b", "i", "a", "strong", "em"]):
    #             tag.unwrap()
    #         wikipedia_text = first_paragraph.get_text(strip=True)

    # ---- Relationships section ----
    h2 = soup.find("h2", class_="relationships")
    if not h2:
        print(f"Warning: No relationships section for {genre_id} – {name}", file=sys.stderr)
        return {
            "id": genre_id,
            "name": name,
            "artistCount": artist_count,
            #"wikipedia": wikipedia_text,
            **{k: [] for k in TARGET_ROWS.values()}
        }

    table = h2.find_next_sibling("table", class_="details")
    if not table:
        print(f"Warning: No details table after relationships for {genre_id} – {name}", file=sys.stderr)
        return {
            "id": genre_id,
            "name": name,
            "artistCount": artist_count,
            #"wikipedia": wikipedia_text,
            **{k: [] for k in TARGET_ROWS.values()}
        }

    result = {
        "id": genre_id,
        "name": name,
        "artistCount": artist_count,
        #"wikipedia": wikipedia_text,
    }

    for row in table.find_all("tr"):
        th = row.find("th")
        td = row.find("td")
        if not th or not td:
            continue
        heading = th.get_text(strip=True).lower()
        if heading in TARGET_ROWS:
            result[TARGET_ROWS[heading]] = _extract_links(td)

    for key in TARGET_ROWS.values():
        result.setdefault(key, [])

    return result



def scrape_genres_from_json(infile: Path, delay: float = 0.25) -> List[Dict[str, Any]]:
    raw = json.loads(infile.read_text())
    genres = raw["genres"]
    results = []

    for i, genre in enumerate(genres, 1):
        result = scrape_single(genre)
        results.append(result)
        print(f"[{i}/{len(genres)}] scraped {genre['name']}")
        time.sleep(delay)

    return results


def _cli():
    import argparse
    p = argparse.ArgumentParser(description="Scrape genre relationship data from MusicBrainz using a JSON input list.")
    p.add_argument("input_json", type=Path, help="Input JSON file with {'genres': [...]} structure")
    p.add_argument("-o", "--output", type=Path, default=Path("output.json"), help="Output JSON file")
    p.add_argument("-d", "--delay", type=float, default=0.5, help="Delay in seconds between requests (default: 0.6)")
    args = p.parse_args()

    results = scrape_genres_from_json(args.input_json, delay=args.delay)
    args.output.write_text(json.dumps(results, indent=2))
    print(f"\nDone. Scraped {len(results)} genres → {args.output}")


if __name__ == "__main__":
    _cli()

File: utils/mbGenresScraper.ts
import * as cheerio from 'cheerio';
import {Genre, MBGenre, SimpleGenre} from '../types';

const BASE_URL = "https://musicbrainz.org/genre/";
const HEADERS = { "User-Agent": "GenreScraper/1.1 (example@example.com)" };

const TARGET_ROWS_LINKS: { [key: string]: GenreRelationKeys } = {
    "subgenre of:": "subgenre_of",
    "subgenres:": "subgenres",
    "has fusion genres:": "fusion_genres",
    "fusion of:": "fusion_of",
    "influenced by:": "influenced_by",
    "influenced genres:": "influenced_genres",
};

const TARGET_ROWS_OTHER: { [key: string]: GenreDataKeys } = {
    "from:": "from",
    "named after area:": "named_after_area",
    "used instruments:": "used_instruments",
};

const UUID_RE = /\/genre\/([0-9a-f\-]{36})/i;

type GenreRelationKeys = {
    [K in keyof Genre]: Genre[K] extends MBGenre[] ? K : never
}[keyof Genre];

type GenreDataKeys = {
    [K in keyof Genre]: K
}[keyof Genre];

function extractLinks($: cheerio.CheerioAPI, cell: cheerio.Cheerio<any>): MBGenre[] {
    const links: MBGenre[] = [];

    cell.find('a[href]').each((_, element) => {
        const href = $(element).attr('href');
        if (!href) return;

        const match = UUID_RE.exec(href);
        if (match) {
            const genreId = match[1];
            const bdi = $(element).find('bdi');
            const name = bdi.length > 0 ? bdi.text().trim() : $(element).text().trim();
            links.push({ id: genreId, name });

        }
    });

    return links;
}

function extractData($: cheerio.CheerioAPI, cell: cheerio.Cheerio<any>): string[] {
    const data: string[] = [];
    cell.find('bdi').each((_, element) => {
        data.push($(element).text().trim());
    });
    console.log(data)
    return data;
}

export async function scrapeSingle(genre: SimpleGenre, genreIds: Set<string>): Promise<Genre> {
    const { id: genreId, name, artistCount } = genre;
    const url = `${BASE_URL}${genreId}`;

    const emptyResult: Genre = {
        id: genreId,
        name,
        artistCount,
        subgenre_of: [],
        subgenres: [],
        fusion_genres: [],
        fusion_of: [],
        influenced_by: [],
        influenced_genres: [],
    };

    try {
        const response = await fetch(url, {
            headers: HEADERS,
            signal: AbortSignal.timeout(30000)
        });

        if (!response.ok) {
            throw new Error(`HTTP ${response.status}`);
        }

        const html = await response.text();
        const $ = cheerio.load(html);

        // Find relationships section
        const relationshipsH2 = $('h2.relationships');
        if (relationshipsH2.length === 0) {
            console.warn(`Warning: No relationships section for ${genreId} – ${name}`);
            return emptyResult;
        }

        const tables = [];
        let table = relationshipsH2.next('table.details');
        if (table.length === 0) {
            console.warn(`Warning: No details table after relationships for ${genreId} – ${name}`);
            return emptyResult;
        }
        while (table.length > 0) {
            tables.push(table);
            table = table.next('table.details');
        }

        const result: Genre = {
            id: genreId,
            name,
            artistCount,
            subgenre_of: [],
            subgenres: [],
            fusion_genres: [],
            fusion_of: [],
            influenced_by: [],
            influenced_genres: [],
            from: [],
            named_after_area: [],
            used_instruments: [],
        };

        for (const details of tables) {
            details.find('tr').each((_, row) => {
                const th = $(row).find('th');
                const td = $(row).find('td');

                if (th.length === 0 || td.length === 0) return;

                const heading = th.text().trim().toLowerCase();
                if (heading in TARGET_ROWS_LINKS) {
                    const key = TARGET_ROWS_LINKS[heading];
                    if (key) result[key] = extractLinks($, td);
                } else if (heading in TARGET_ROWS_OTHER) {
                    const key = TARGET_ROWS_OTHER[heading];
                    if (key) { // @ts-ignore
                        result[key] = extractData($, td);
                    }
                }
            });
        }


        return result;

    } catch (error) {
        console.error(`Error fetching ${genreId} – ${name}:`, error);
        return emptyResult;
    }
}

export async function scrapeGenres(genres: SimpleGenre[], delayMs: number = 500): Promise<Genre[]> {
    const genreIds = new Set(genres.map(g => g.id));
    const results: Genre[] = [];

    for (let i = 0; i < genres.length; i++) {
        const genre = genres[i];
        const result = await scrapeSingle(genre, genreIds);
        results.push(result);
        console.log(`[${i + 1}/${genres.length}] scraped ${genre.name}`);

        if (i < genres.length - 1) {
            await new Promise(resolve => setTimeout(resolve, delayMs));
        }
    }

    return results;
}